### DISCUSSION TITLE
Unable to communicate using the Hello World application

### DISCUSSION THREAD

**Discussion by ted-miller** [Yaskawa-Global/YMConnect#7] [Q&A] [answered]
This was originally in another discussion: https://github.com/Yaskawa-Global/YMConnect/discussions/4

@RubyAI132 writes:

> Hello
> I get a failure when trying to run the Hello world (quick start ) project
> I didn't made any setting adjustment to the controller. Do I need do sett something in the controller?
> I have a YRC1000 micro controller and I'm using visual studio 2022 to compile and run/ debug the solution
> I changed the ip address parameter in OpenConnection to the ip of the controlee and compiled the sample code successfully
> I made sure I have ping to the controller ip and started debug .
> 
> The issue I see is that there is no response from the controller
> This is the message I see on the console
> 'Code (196614): No response from controller after 3 retries.'
> 
> The code line that fails is this
> code line 1378-1380
> // proxy allocated, drain it
> for (auto _Pnext = _STD exchange(_Myproxy->_Myfirstiter, nullptr); _Pnext; _Pnext = _Pnext->_Mynextiter) {
> _Pnext->_Myproxy = nullptr;

@aseamans:
> Today, we went ahead and created a page for instructions to [set up the robot controller for YMConnect.](https://developer.motoman.com/en/YMConnect/RobotControllerSetup)

@ted-miller:
> the message Code (196614): No response from controller after 3 retries. indicates that the PC was unable to connect to the robot controller. This is most likely due to a misconfigured parameter. The link from aseamans provides the required settings.

> When OpenConnection fails to connect to the robot controller, it returns a null pointer. The failure on code line 1378-1380 is most likely due to trying to use the null pointer.

@RubyAI132:

> I followed the controller setup guide as well as the Motoman ethernet function manual
> It turned out my controller factory setting had the High Speed Ethernet function enabled .
> But still I can't establish connection for simple hello world
> 
> I did packet capture with Wireshark on my PC (windows10) network interface while trying to run the quick start code and I observed no! packet were sent out from the PC with the controller address .
> 
> I made sure firewalls are disabled and that there is ping to the controller but no success to run Hello World
> Do I need to look / try something else
> 
> I also use the controller High speed ethernet error log utility  and don’t see any activity
> 
> The [attached archive](https://github.com/Yaskawa-Global/YMConnect/files/14669255/YMCONNECT.zip)  has  the COMS.BIN and the packet dump
>  I send ping prior to starting the YMC Hello World
> The dump period includes both the ping and the YMC Hello World until failing
> 
> The controller ip is 192.168.100.1
> And PC is 192.168.100.77

**Comment by ted-miller** [comment]
Hello @RubyAI132

Using your CMOS.BIN, we were able to connect to the robot controller. So, your robot is configured properly for communication. We believe the issue is on your PC.

Please try adding an explicit exception in your firewall for UDP 10040 and 10041. 
- In the START menu, search for `Windows Defender Firewall with Advanced Security`
- Select `Outbound Rules` and click `New Rule`
- Select `Port`, then click `Next`
- Select `UDP`, then `Specific remote ports`. Enter `10040-10041`, then click `Next`
- Select `Allow the connection`, then click `Next`
- Check all three options, `Domain`, `Private`, `Public`, then click Next
- Give it a name, such as YMConnect, then click `Finish`

Additionally, if using the Hello World example verbatim, please check that the controller IP address is specified correctly as `192.168.100.1`.

![image](https://github.com/Yaskawa-Global/YMConnect/assets/4623811/ce7daae0-2b25-44d9-894f-f53d21e2f692)

Lastly, if your computer has multiple network adapters, make sure that they don't have overlapping subnets. So, only one adapter should be able to access the `192.168.100.xxx` range of addresses.

**Reply by ted-miller** [reply]
One more thing to verify is that your executable and the DLL must both be configured for `release` or `debug`. It will fail if attempting to run a debug-executable with a release-dll (and vice versa). 

I don't think this is a problem in your case, because you're getting a valid error code returned. But it's still worth checking.

I'll be updating the quick-start guide to reflect both configurations.

**Comment by ted-miller** [comment]
Closing this discussion. Please reopen if the problem persists.

**Comment by yfc26** [comment]
I have the same problem. "  // proxy allocated, drain it
    for (auto _Pnext = _STD exchange(_Myproxy->_Myfirstiter, nullptr); _Pnext; _Pnext = _Pnext->_Mynextiter) {
        _Pnext->_Myproxy = nullptr;
    }" in xmemory. It is from myexe.exe!StatusInfo::~StatusInfo(). It seems that this is caused by ~StatusInfo funciton. please check if it has a return in Switch Default condition.

**Reply by aseamans** [reply]
Most of the time, this error is caused by an unsuccessful connection to the controller. Are you able to ping the controller? Also, check your PC network settings as seen in the above answer from Ted on 21 March. In addition, verify that you are using the debug version of the DLL with the debug version of your application. 
I'll reopen this discussion, let us know if this issue still isn't solved.
Thank you.

**Reply by yfc26** [reply, marked as answer]
adding "return status" solve the problem. thank you.
